engine/src/main/java/org/terasology/math/Diamond3iIterable.java

```diff
@@ -0,0 +1,162 @@
+ * the shape can be an hull with distance 1 to a full solid.
+ */
+public final class Diamond3iIterable implements Iterable<Vector3ic> {
+
+    private final Vector3i origin = new Vector3i();
+    private final int endDistance;
+    private int startDistance = 0;
+
+    /**
+     *
+     * @param origin center point of Diamond3iIterator
+     * @param maxDistance maxDistance away
+     */
+    private Diamond3iIterable(Vector3ic origin, int maxDistance) {
+        Preconditions.checkArgument(maxDistance >= 0, "maxDistance must be > 0");
+        this.origin.set(origin);
+        this.endDistance = maxDistance;
+    }
+
+    /**
+     *
+     * @param origin center point of Diamond3iIterator
+     * @param startDistance inner distance outwards
+     * @param maxDistance maxDistance away
+     */
+    private Diamond3iIterable(Vector3ic origin, int startDistance, int maxDistance) {
+        Preconditions.checkArgument(startDistance < maxDistance, "startDistance must be < maxDistance");
+        Preconditions.checkArgument(maxDistance >= 0, "maxDistance must be >= 0");
+        Preconditions.checkArgument(startDistance >= 0, "maxDistance must be >= 0");
+
+        this.origin.set(origin);
+        this.endDistance = maxDistance;
+        this.startDistance = startDistance;
+    }
+
+    /**
+     * solid diamond shape from the center to a radius
+     * @param origin center of diamond iterator
+     * @param radius distance to iterate out to
+     */
+    public static Diamond3iIterable.Builder region(Vector3ic origin, int radius) {
+        return new Diamond3iIterable.Builder(origin, radius);
+    }
+
+    /**
+     * hollow shape with an inner and outer distance to iterate over
+     * @param origin  center of diamond iterator
+     * @param start distance to start iterating from
+     * @param end distance to iterate out to
+     */
+    public static Diamond3iIterable.Builder hollow(Vector3ic origin, int start, int end) {
+        return new Diamond3iIterable.Builder(origin, end).start(start);
+    }
+
+    /**
+     * iterates in a shell that is 1 block wide
+     * @param origin center of shell
+     * @param radius distance for 1 block wide shell
+     */
+    public static Diamond3iIterable.Builder shell(Vector3ic origin, int radius) {
+        return new Diamond3iIterable.Builder(origin, radius).start(radius - 1);
+    }
+
+    @Override
+    public Iterator<Vector3ic> iterator() {
+        Vector3i pos = new Vector3i();
+        final int[] level = {this.startDistance};
+        final int[] offset = {-this.startDistance,0,0};
+
+        return new Iterator<Vector3ic>() {
+            @Override
+            public boolean hasNext() {
+                return level[0] < endDistance;
+            }
+
+            @Override
+            public Vector3ic next() {
+                pos.set(origin.x + offset[0], origin.y + offset[1], origin.z + offset[2]);
```

Comment:
If a `Vector3i` was used no represent the offset, the statement could be simplified to:
`origin.add(offset, pos);`